In [3]:
# import packages
import pandas as pd

## 1) Load the messy dataset (with headers)

In [4]:
# messy_employee_health.csv
url = "https://raw.githubusercontent.com/bkeerthana/MLC/refs/heads/main/messy_employee_health.csv"
df = pd.read_csv(url)

## 2) Missing headers (use the no-header file)


In [5]:
# check the first first records
df.head()

ID  Full Name  Age    GENDER Heart Rate  Height  \
0    1.0       Anu  22.0   Female     72 bpm  170 cm   
1    2.0     Rahul   NaN     MALE      85bpm  1.75 m   
2    3.0     Meera  29.0  female         NaN  175 cm   
3    4.0      Sidd  -5.0     Male     90 bpm   180cm   
4    5.0     Priya  31.0   FEMALE    100 BPM  1.60 m   

                                  Address       Join Date       Notes  \
0           12 MG Road, Bengaluru, 560001      2024-01-05        good   
1             45 Park St, Kolkata, 700016      05/02/2024  naïve café   
2          77 Anna Salai, Chennai, 600002      2024/03/10         NaN   
3             1 Marine Dr, Mumbai, 400001  March 15, 2024           ?   
4  No. 9, Residency Rd, Bengaluru, 560025      2024-13-01         NaN   

  Constant Column Irrelevant Column  Empty Column  
0            SAME                x1           NaN  
1            SAME                x2           NaN  
2            SAME                x3           NaN  
3            SAME                x4           NaN  
4            SAME                x5           NaN

## 3) Missing values
Compute missing counts and percentages. Then choose a strategy (drop, fill, impute).

In [ ]:
# drop the missing values

In [ ]:
# fill the missing values with average

In [ ]:
# impute the values

## 4) Empty rows and empty columns
Identify and remove completely empty rows and columns.

In [ ]:
# TODO: find empty rows

In [ ]:
# TODO: drop empty rows

In [ ]:
# TODO: find empty columns

In [ ]:
# TODO: drop empty columns

In [ ]:
# df_step.shape

## 5) Messy column names
Standardize: strip, lowercase, underscores, remove special chars.

In [ ]:
# TODO: clean column names




## 6) Multiple variables in one column (Heart Rate)
Extract numeric heart rate (bpm) as an integer. Handle values like `85bpm`, `100 BPM`, `??`.

In [ ]:
# TODO: normalize heart_rate strings then extract digits

In [ ]:

# df_step['heart_rate'] = ...
# df_step['heart_rate_bpm'] = ...

# df_step[['heart_rate', 'heart_rate_bpm']].head(10)

## 7) Inconsistent units (Height)
Convert height to meters (float). Handle `170 cm`, `1.75 m`, `180cm`.

In [ ]:
# TODO: create height_m
# df_step['height_m'] = ...

# df_step[['height', 'height_m']].head(10)

## 8) Capitalization + typos in categorical values (Gender)
Standardize gender values and fix typos like `Mle`.

In [ ]:
# TODO: normalize gender
# df_step['gender'] = ...

# df_step['gender'].value_counts(dropna=False)

## 9) Data types (Age, IDs)
Convert `id` to integer (nullable ok), `age` to numeric; identify invalid values (e.g., negative age).

In [ ]:
# TODO: convert types
# df_step['id'] = ...
# df_step['age'] = ...

# TODO: find invalid ages
# df_step[(df_step['age'] < 0) | (df_step['age'] > 100)]

## 10) Addresses
Split address into street, city, and pincode. Handle missing parts safely.

In [ ]:
# TODO: split address by comma into 3 parts
# addr = df_step['address'].str.split(',', n=2, expand=True)
# df_step['street'] = ...
# df_step['city'] = ...
# df_step['pincode'] = ...

# df_step[['address','street','city','pincode']].head(10)

## 11) Dates
Parse join_date with mixed formats and coerce invalid dates to NaT.

In [ ]:
# TODO: parse dates
# df_step['join_date'] = pd.to_datetime(df_step['join_date'], errors='coerce', dayfirst=True)
# df_step[['join_date']].head(10)

## 12) Non-ASCII characters (Notes)
Remove non-ASCII characters OR normalize them.

In [ ]:
# TODO: remove non-ascii
# df_step['notes_clean'] = ...
# df_step[['notes','notes_clean']].head(10)

## 13) Unnecessary data
Remove duplicate rows, irrelevant columns, and uninformative columns.

In [ ]:
# TODO: duplicates
# df_step.duplicated().sum()
# df_step = df_step.drop_duplicates()

# TODO: drop irrelevant columns
# df_step = df_step.drop(columns=['irrelevant_column'], errors='ignore')

# TODO: drop uninformative columns (nunique == 1)
# nun = df_step.nunique(dropna=False)
# drop_cols = nun[nun == 1].index
# df_step = df_step.drop(columns=drop_cols)

# df_step.shape

## 14) Final cleaned dataset
Show summary and export as CSV.